CARET_trace/src/hooked_trace_points.cpp

```diff
@@ -233,6 +237,29 @@ int dds_write_impl(void * wr, void * data, long tstamp, int action)  // NOLINT
   return dds_return;
 }
 
+// for cyclonedds
+// bind : &ros_message -> source_timestamp
+// cspell: ignore ddsi, serdata, dinp
+int dds_writecdr_impl(void * wr, void * xp, struct ddsi_serdata * dinp, bool flush)  // NOLINT
+{
+  static auto & context = Singleton<Context>::get_instance();
+  using functionT = int (*)(void *, void *, struct ddsi_serdata *, bool);  // NOLINT
+
+  // clang-format on
+  if (CYCLONEDDS::DDS_WRITECDR_IMPL == nullptr) {
+    update_dds_function_addr();
+  }
+  int dds_return = ((functionT)CYCLONEDDS::DDS_WRITECDR_IMPL)(wr, xp, dinp, flush);
+
+  if (context.is_recording_allowed()) {
+    tracepoint(TRACEPOINT_PROVIDER, dds_bind_addr_to_stamp, dinp, dinp->timestamp.v);
```

Comment:
I think `message` argument (`dinp`) of this `dds_bind_addr_to_stamp` is not equal to `message` argument of `rclcpp_publish` by [rclcpp PR #7](https://github.com/tier4/rclcpp/pull/7). Can you make them equal? I think the following method works.
- hook `rmw_publish_serialized_message` and save `serialized_message` to thread local storage
- use saved `sereialized_message` here as `message` argument of `dds_bind_addr_to_stamp`

Also, I think [rclcpp PR #7](https://github.com/tier4/rclcpp/pull/7) need to be changed. (change `message` argument of `rclcpp_publish` from `&message` to `&message.get_rcl_serialized_message()`)

Reply:
By using thread-local memory, I have modify it so that the message address for `rclcpp_publish` matches the message address for `dds_bind_addr_to_stamp`.

Fixed in following links.
- caret_trace
  - [820baf9](https://github.com/tier4/caret_trace/pull/228/commits/820baf92c0e0e831702350615038b6a86997b235).
- rclcpp
  - [7d478c4](https://github.com/tier4/rclcpp/pull/7/commits/7d478c4a73204fd57a4511241c197b10d57bce99)